workflow/controller/workflowpod.go

```diff
@@ -622,6 +628,41 @@ func (woc *wfOperationCtx) addInputArtifactsVolumes(pod *apiv1.Pod, tmpl *wfv1.T
 	return nil
 }
 
+// addOutputArtifactsVolumes mirrors any volume mounts in the main container to the wait sidecar.
+// For any output artifacts that were produced in mounted volumes (e.g. PVCs, emptyDirs), the
+// wait container will collect the artifacts directly from volumeMount instead of `docker cp`-ing
+// them to the wait sidecar. In order for this to work, we mirror all volume mounts in the main
+// container under a well-known path.
+func addOutputArtifactsVolumes(pod *apiv1.Pod, tmpl *wfv1.Template) {
+	if tmpl.GetType() == wfv1.TemplateTypeResource {
+		return
+	}
+	mainCtrIndex := -1
+	waitCtrIndex := -1
+	var mainCtr *apiv1.Container
+	for i, ctr := range pod.Spec.Containers {
+		switch ctr.Name {
+		case common.MainContainerName:
+			mainCtrIndex = i
+		case common.WaitContainerName:
+			waitCtrIndex = i
+		}
+	}
+	if mainCtrIndex == -1 || waitCtrIndex == -1 {
+		panic("Could not find main or wait container in pod spec")
+	}
+	mainCtr = &pod.Spec.Containers[mainCtrIndex]
+	waitCtr := &pod.Spec.Containers[waitCtrIndex]
+
+	for _, mnt := range mainCtr.VolumeMounts {
+		mnt.MountPath = filepath.Join(common.ExecutorMainFilesystemDir, mnt.MountPath)
+		// ReadOnly is needed to be false for overlapping volume mounts
+		mnt.ReadOnly = false
```

Comment:
This clashes with PSPs using read-only `allowedHostPaths`.

If a workflow container has overlapping volume mounts, will it not have to disable readOnly itself and won't `mnt.ReadOnly` already be set to the right value?